### DISCUSSION TITLE
Unhandled rejection caught by vitest (I think I'm going mad)

### DISCUSSION THREAD

**Discussion by Ancient-Dragon** [vitest-dev/vitest#1649] [General]
Hi there, I cannot seem to figure out why vitest seems to think it is handling an uncaught rejection. The idea of this test is that the setAttributesMock should throw an exception, I'm calling await expect(...).rejected etc and all the tests are passing. Any help would be greatly appreciated

My real code looks something like:

```typescript
const traceFunction = async <T>(callback: Function, options?: TraceFunctionOptions): Promise<T> => {
    let functionError: unknown;
    let otelError: unknown;
    const functionName = callback.name === '' ? `${route?.path} - Unknown Function` : callback.name;
    const defaultedOptions = {
      attributes: { ...defaultAttributes },
      retryWithoutOpenTelemetry: false,
      ...options,
    };

    const callbackResult: T = await validateOpenTelSetup().tracer.startActiveSpan(functionName, async (sp: Span) => {
      try {
        sp.setAttributes(defaultedOptions.attributes);
        try {
          return await callback();
        } catch (error) {
          functionError = error;
          return Promise.resolve();
        }
      } catch (error) {
        if (defaultedOptions.retryWithoutOpenTelemetry) {
          return await callback();
        }
        otelError = error;
        return Promise.reject(error);
      } finally {
        sp.end();
      }
    });
    if (functionError !== undefined) {
      return Promise.reject(functionError);
    }

    if (otelError !== undefined) {
      return Promise.reject(otelError);
    }
    return Promise.resolve(callbackResult);
  };

```

My test code looks something like this: 
```typescript
const setAttributesMock = vi.fn().mockImplementation((attributes) => {
  if (Object.keys(attributes).includes('testFailure')) {
    throw new Error('Opentelemetry failure');
  }
});

const spanMock = {
  end: vi.fn(),
  setAttributes: setAttributesMock,
};

const startActiveSpanMock = vi.fn().mockImplementation((spanName, callback) => {
  callback(spanMock);
});
// eslint-disable-next-line
import { useOpenTelemetry } from './OpenTelemetry'

vi.mock('custom-package', () => {
  const mockedObject: OpenTelemetryTracerAndLogger = {
    logger: vi.mock as any,
    tracer: {
      startActiveSpan: startActiveSpanMock,
      startSpan: vi.fn(),
    },
    initialiseInstrumentation: vi.fn(),
  };
  const initOpenTelemetryMocked = vi.fn().mockResolvedValue(mockedObject);

  return { initOpenTelemetry: initOpenTelemetryMocked };
});

test('successfully fails the function is open-telemetry fails', async () => {
    const { traceFunction, setupOpenTelemetry } = useOpenTelemetry();
    await setupOpenTelemetry();

    const callback = vi.fn().mockImplementation(() => {
      console.log('Mock Impl');
    });
    await expect(() => traceFunction(callback, {
      attributes: {
        testFailure: 'test',
      },
    })).rejects.toThrow('Opentelemetry failure');

    expect(startActiveSpanMock).toBeCalledTimes(1);
    expect(spanMock.end).toBeCalledTimes(1);
    expect(callback).toBeCalledTimes(0);
    expect(startActiveSpanMock.mock.calls[0][0]).toEqual('spy');
  });
```

I know it probably isn't a false positive but there seems to be no way for me to ignore this warning / error. If I put a try catch around the whole test it still fails the tests command... does anyone know how I can disable this warning (as a last resort)

**Comment by sheremet-va** [comment]
Unhandled rejection means you are trowing an error outside of test, and Vitest couldn't track it. Also, I doubt you `vi.mock` actually works, because it is hoisted to top of the file, so you don't have access to defined variables.

**Comment by Ancient-Dragon** [comment]
So because I'm mocking method so that it will throw an error and it is defined outside of my test vitest will fail? What is the best way to approach mocking a library, so that it throws an error?

It seems work, I was able to catch the correct error in my code and handle it as I wanted, the error also gets passed out as you can see in the expects().rejected (the hacky bit here is that I import just before calling vi.mock which stops it being hosited above the variables.

**Comment by sheremet-va** [comment]
> So because I'm mocking method so that it will throw an error and it is defined outside of my test vitest will fail? What is the best way to approach mocking a library, so that it throws an error?

No, you have a code that runs after the test. This might happened because of not cleared timer, or not awaited promise. Try reproducing error on stackblitz.

**Reply by Ancient-Dragon** [reply]
Is there any way to see what code is running outside of the test? When I tried using logging I cannot see it re-throwing any error and the last log line is the one I put at the end of the test.

**Reply by Ancient-Dragon** [reply]
So I added logging to all my catch blocks, as well as to the place I throw the exception and the end of the test, this is what it looks like: 

```
2022-07-17T17:07:11.809Z Throwing error
2022-07-17T17:07:11.810Z Experience open-telemetry error Error: Opentelemetry failure...
2022-07-17T17:07:11.812Z Otel Error - finishing up
2022-07-17T17:07:11.814Z Finished test

```
